index.js

```diff
@@ -702,20 +702,93 @@
     }
 
     function categorizeCommitMessage(message) {
+        message = message.trim();
         const categoryResult = categoryRegex.exec(message);
         if (!categoryResult)
             return ["Uncategorized"];
 
         const categoryString = categoryResult[1];
         let categories = [];
         if (categoryString.indexOf("+") >= 0)
-            categories.push(...categoryString.split("+"));
+            categories.push(...categoryString.split("+").map(c => c.trim()));
         else
-            categories.push(categoryString);
+            categories.push(categoryString.trim());
 
         return categories;
     }
 
+    class Category extends Array {
+        #commits = [];
+        #subCategories = new Map();
+
+        constructor(superCategory, name) {
+            super();
+            this.superCategory = superCategory;
+            this.name = name;
+        }
+
+        get validSelector() {
+            let validSelectorCategory = this.name.replace(invalidSelectorCharacters, "");
+            if (startsWithNumberRegex.test(validSelectorCategory))
+                // Selectors starting with a number are invalid. Just prepend an 'i' to counteract it.
+                validSelectorCategory = "i" + validSelectorCategory;
+            return this.isRoot ? validSelectorCategory : `${this.superCategory.validSelector}-${validSelectorCategory}`;
+        }
+
+        get isRoot() {
+            return !this.superCategory;
+        }
+
+        get commitCount() {
+            return this.#commits.length;
+        }
+
+        // The given categories are just strings.
+        // Inserts the commit into the lowest-level category for each of the categories given.
+        insertInto(commit, categories) {
+            for (const category of categories) {
+                const lcCategory = category.toLowerCase();
+                if (lcCategory.indexOf("/") < 0) {
+                    if (!this.#subCategories.has(lcCategory)) this.#subCategories.set(lcCategory, new Category(this, category));
+                    this.#subCategories.get(lcCategory).#commits.push(commit);
+                } else {
+                    const [firstSubCategory, lowerCategories] = category.split("/", 2);
+                    if (!this.#subCategories.has(firstSubCategory.toLowerCase()))
+                        this.#subCategories.set(firstSubCategory.toLowerCase(), new Category(this, firstSubCategory));
+                    this.#subCategories.get(firstSubCategory.toLowerCase()).insertInto(commit, [lowerCategories]);
+                }
+            }
+        }
+
+        sortedSubcategories() {
+            // For the sort: https://stackoverflow.com/a/45544166
+            const allKeys = [];
+            for (let k of this.#subCategories.keys()) allKeys.push(k);
```

Comment:
This can be achieved via the spread operator.
```suggestion
            const allKeys = [...this.#subCategories.keys()];
```

Reply:
Did I mention my JS was rusty?